contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -374,6 +569,189 @@ process_remove_all(struct sshbuf* request, struct sshbuf* response, struct agent
 	return r;
 }
 
+#ifdef ENABLE_PKCS11
+int process_add_smartcard_key(struct sshbuf* request, struct sshbuf* response, struct agent_connection* con)
+{
+	char *provider = NULL, *pin = NULL, canonical_provider[PATH_MAX];
+	int i, count = 0, r = 0, request_invalid = 0, success = 0;
+	struct sshkey **keys = NULL;
+	struct sshkey* key = NULL;
+	size_t pubkey_blob_len, provider_len, pin_len, epin_len;
+	u_char *pubkey_blob = NULL;
+	char *thumbprint = NULL;
+	char *epin = NULL;
+	HKEY reg = 0, sub = 0, user_root = 0;
+	SECURITY_ATTRIBUTES sa = { 0, NULL, 0 };
+
+	pkcs11_init(0);
+
+	if ((r = sshbuf_get_cstring(request, &provider, &provider_len)) != 0 ||
+		(r = sshbuf_get_cstring(request, &pin, &pin_len)) != 0 ||
+		pin_len > 256) {
+		debug("add smartcard request is invalid");
+		request_invalid = 1;
+		goto done;
+	}
+
+	if (realpath(provider, canonical_provider) == NULL) {
+		debug("failed PKCS#11 add of \"%.100s\": realpath: %s",
+			provider, strerror(errno));
+		request_invalid = 1;
+		goto done;
+	}
+
+	// Remove 'drive root' if exists
+	if (canonical_provider[0] == '/')
+		memmove(canonical_provider, canonical_provider + 1, strlen(canonical_provider));
+	if (get_user_root(con, &user_root) != 0 ||
+		is_reg_sub_key_exists(user_root, SSH_PKCS11_PROVIDERS_ROOT, canonical_provider))
```

Comment:
If a customer has added PKCS11 to registry then the current logic returns an error "ssh-agent refused connection" to ssh-add. 

The ideal user experience is to override the existing keys.
We should remove the existing keys and proceed with the current flow.

Reply:
instead of goto done at line 608.. It should be

remove_matching_subkeys_from_registry(user_root, SSH_KEYS_ROOT, L"comment", canonical_provider);
		remove_matching_subkeys_from_registry(user_root, SSH_PKCS11_PROVIDERS_ROOT, L"provider", canonical_provider);

Reply:
This is done intentionally to mimic the exact behavior of [openssh/openssh-portable](https://github.com/openssh/openssh-portable). I've addressed that in the discussion [here](https://github.com/PowerShell/openssh-portable/pull/537#discussion_r763754974).